File: selenium-python/pages/base_page.py
from __future__ import annotations

from selenium.common.exceptions import NoSuchElementException
from selenium.webdriver.common.by import By
from selenium.webdriver.remote.webdriver import WebDriver
from selenium.webdriver.remote.webelement import WebElement
from selenium.webdriver.support.ui import WebDriverWait


class BasePage:
    def __init__(self, driver: WebDriver, timeout: int = 5) -> None:
        self.driver = driver
        self.timeout = timeout

    def by_test_id(self, test_id: str) -> tuple[str, str]:
        return By.CSS_SELECTOR, f'[data-testid="{test_id}"]'

    def wait_for_test_id(self, test_id: str) -> WebElement:
        locator = self.by_test_id(test_id)
        return WebDriverWait(self.driver, self.timeout).until(
            lambda drv: drv.find_element(*locator)
        )

    def click_on(self, test_id: str) -> None:
        self.wait_for_test_id(test_id).click()

    def type_into(self, test_id: str, value: str) -> None:
        element = self.wait_for_test_id(test_id)
        element.clear()
        element.send_keys(value)

    def get_test_id_text(self, test_id: str) -> str:
        return self.wait_for_test_id(test_id).text

    def is_test_id_visible(self, test_id: str) -> bool:
        try:
            return self.driver.find_element(*self.by_test_id(test_id)).is_displayed()
        except NoSuchElementException:
            return False

File: selenium-java/src/test/java/frameworks/seleniumjava/pages/BasePage.java
package frameworks.seleniumjava.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.WebDriverWait;
import io.qameta.allure.Step;

import java.time.Duration;

public abstract class BasePage {
  protected final WebDriver driver;
  private final Duration timeout;

  protected BasePage(WebDriver driver) {
    this(driver, Duration.ofSeconds(5));
  }

  protected BasePage(WebDriver driver, Duration timeout) {
    this.driver = driver;
    this.timeout = timeout;
  }

  @Step
  protected By byTestId(String testId) {
    return By.cssSelector("[data-testid='" + testId + "']");
  }

  @Step
  protected WebElement waitForTestId(String testId) {
    By locator = byTestId(testId);
    return new WebDriverWait(driver, timeout).until(browser -> browser.findElement(locator));
  }

  @Step
  protected void clickOn(String testId) {
    waitForTestId(testId).click();
  }

  @Step
  protected void typeInto(String testId, String value) {
    WebElement element = waitForTestId(testId);
    element.clear();
    element.sendKeys(value);
  }

  @Step
  protected String getTestIdText(String testId) {
    return waitForTestId(testId).getText();
  }

  @Step
  protected boolean isTestIdVisible(String testId) {
    try {
      return driver.findElement(byTestId(testId)).isDisplayed();
    } catch (NoSuchElementException error) {
      return false;
    }
  }
}
